### PR TITLE
Finish seg train loop

### PR DESCRIPTION
**Pull request by guarin**
## What has changed and why?

* Finishes first version of seg train loop
* Added some bugfixes

## How has it been tested?

* Manually
* Unit tests

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_commands/train_task_helpers.py [hunk 1]
```diff
@@ -29,40 +31,58 @@
...
+from lightly_train._transforms.task_transform import TaskTransform
 from lightly_train.types import PathLike, TaskDatasetItem
 
 logger = logging.getLogger(__name__)
 
 
 @contextmanager
-def rank_zero_unshared_only(
+def filesystem_rank_zero_first(
     fabric: Fabric, path: PathLike
```
Comment (guarin) [inline src/lightly_train/_commands/train_task_helpers.py:52]: Had to rework this because it resulted in a deadlock with DDP.

File: src/lightly_train/_transforms/semantic_segmentation_transform.py [hunk 2]
```diff
@@ -79,7 +79,7 @@ def __init__(self, transform_args: SemanticSegmentationTransformArgs) -> None:
                     pad_if_needed=transform_args.random_crop.pad_if_needed,
                     pad_position=transform_args.random_crop.pad_position,
                     fill=transform_args.random_crop.fill,
-                    mask_value=transform_args.random_crop.fill_mask,
+                    fill_mask=transform_args.random_crop.fill_mask,
```
Comment (guarin) [inline src/lightly_train/_transforms/semantic_segmentation_transform.py:82]: `mask_value` doesn't exist

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py [hunk 3]
```diff
@@ -30,29 +36,64 @@ class DINOv2SemanticSegmentationTrainArgs(TaskTrainModelArgs):
...
+        self.train_miou.update(pred_masks, masks)
+        return TaskStepResult(
+            loss=loss,
+            log_dict={
+                "train_loss": loss.detach(),
+                "train_metric/miou": self.train_miou,
+            },
+        )
 
     def validation_step(
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py:81]: In the main training loop `validation_step` is placed under the `torch.no_grad` context. Should we add the `torch.no_grad` decorator in case it is called from somewhere else?

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py [hunk 4]
```diff
@@ -30,29 +36,64 @@ class DINOv2SemanticSegmentationTrainArgs(TaskTrainModelArgs):
 
 class DINOv2SemanticSegmentationTrain(TaskTrainModel):
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py:37]: Should we have a method to save the model, optimizer, etc? Or is everything already handled by fabric?

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py [hunk 5]
```diff
@@ -59,13 +59,20 @@ def __init__(
             backbone_weights_path = model_args.pop("backbone_weights")
 
         # Get the backbone.
+        # TODO(Guarin, 07/25): This builds a student model. Double check that this is
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py]: The teacher is used for evals in general.
If we load teacher weights, would there still be a difference afterwards? Mask tokens?

**Review by stegmuel — APPROVED** [review]
LGTM!

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py [hunk 6]
```diff
@@ -30,29 +36,64 @@ class DINOv2SemanticSegmentationTrainArgs(TaskTrainModelArgs):
...
         )
-        self.criterion: Module
-        self.metric: Module
+        self.criterion = DINOv2SemanticSegmentationCrossEntropyLoss()
+        self.val_loss = MeanMetric()
+
+        # MeanIoU assumes that background is class 0.
+        self.train_miou = MeanIoU(
+            num_classes=max(data_args.classes) + 1,
+            include_background=True,
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py:59]: Can you put a TODO  to make the `include_background` settable ?

**Review by stegmuel — APPROVED** [review]

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py (hunk 3 above)
Comment (guarin) [reply src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py:81]: Hmm, IMO it is the job of the outside code to call `model.eval()` and `torch.no_grad`. Otherwise we would have to call it each time in the `validation_step`.

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py (hunk 4 above)
Comment (guarin) [reply src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py:37]: No we have to implement this. Would implement it in the main loop though.

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py (hunk 5 above)
Comment (guarin) [reply src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py]: The difference between teacher/student is the drop path rate. I updated the code to set it to 0 by default.

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py (hunk 4 above)
Comment (guarin) [reply src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation_train.py:37]: I'll address this in a follow-up PR. Together with the backbone weights loading.

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py [hunk 7]
```diff
@@ -58,14 +58,26 @@ def __init__(
         if model_args is not None and "backbone_weights" in model_args:
             backbone_weights_path = model_args.pop("backbone_weights")
 
+        # Disable drop path by default.
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py:61]: I would keep stochastic depth as settable.
It seems particularly useful for large models and can accelerate training if properly implemented.
See Table 13 from [DeiT III](https://hal.science/hal-03945731/document)

Comment (guarin) [reply src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py:61]: Will add option to keep it configurable in a follow-up 👍🏼
